src/runtime/local/kernels/EigenCal.h

```diff
@@ -0,0 +1,157 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SRC_RUNTIME_LOCAL_KERNELS_EIGEN_H
+#define SRC_RUNTIME_LOCAL_KERNELS_EIGEN_H
+
+#pragma once
+
+#include <runtime/local/context/DaphneContext.h>
+#include <runtime/local/datastructures/CSRMatrix.h>
+#include <runtime/local/datastructures/DataObjectFactory.h>
+#include <runtime/local/datastructures/DenseMatrix.h>
+
+#include <Eigen/Eigenvalues> // header file
+#include <Eigen/Dense>
+
+#include <iostream>
+
+
+#include <cassert>
+#include <cstddef>
+
+
+// ****************************************************************************
+// Struct for partial template specialization
+// res1 matrix for eigenvalues, res2 matrix for eigenvectors
+// Column k of the returned matrix res2 is an eigenvector corresponding to eigenvalue number k as returned by eigenvalues(). 
+// The eigenvectors are normalized to have (Euclidean) norm equal to one. 
+// ****************************************************************************
+
+template<class DTRes1, class DTRes2, class VTArg>
+struct EigenCal {
+    static void apply(DTRes1 *& res1, DTRes2 *& res2, const VTArg * inMat, DCTX(ctx)) = delete;
+};
+
+// ****************************************************************************
+// Convenience function
+// ****************************************************************************
+
+
+template<class DTRes1, class DTRes2, class VTArg>
+void eigenCal(DTRes1 *& res1, DTRes2 *& res2, const VTArg * inMat, DCTX(ctx)) {
+    EigenCal<DTRes1, DTRes2, VTArg>::apply(res1,res2, inMat, ctx);
+}
+
+// ****************************************************************************
+// (Partial) template specializations for different data/value types
+// ****************************************************************************
+
+// ----------------------------------------------------------------------------
+// DenseMatrix 
+// Double Value types as input
+// ----------------------------------------------------------------------------
+
+template<>
+struct EigenCal<DenseMatrix<double>,DenseMatrix<double>,DenseMatrix<double>> {
+    static void apply(DenseMatrix<double> *& res1,DenseMatrix<double> *& res2, const DenseMatrix<double> * inMat, DCTX(ctx)) {
+        const auto nr = static_cast<size_t>(inMat->getNumRows());
+        const auto nc = static_cast<size_t>(inMat->getNumCols());
+        Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> inputMatrix(nr, nc);
+       
+       std::tuple<int, int>{1, -1};
+        for(size_t r=0; r<nr; r++){
+            for(size_t c=0; c<nc; c++){
+                inputMatrix(r,c)=inMat->get(r,c);
+            }
+        }
```

Comment:
Please try to avoid these copy loops. Eigen can work on array data. Maybe this helps to get you started:
https://stackoverflow.com/questions/28722899/creating-an-eigen-matrix-from-an-array-with-row-major-order